### PR TITLE
7903464: JMH: Add virtual threads executor

### PR DESCRIPTION
**Pull request by kuksenko**
The list of working threads executors should be expanded with virtual threads executor.
Added VIRTUAL_TPE which is  "Executors.newFixedThreadPool(maxThreads, virtualThreadFactory)).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [CODETOOLS-7903464](https://bugs.openjdk.org/browse/CODETOOLS-7903464): JMH: Add virtual threads executor

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jmh.git pull/102/head:pull/102` \
`$ git checkout pull/102`

Update a local copy of the PR: \
`$ git checkout pull/102` \
`$ git pull https://git.openjdk.org/jmh.git pull/102/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 102`

View PR using the GUI difftool: \
`$ git pr show -t 102`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jmh/pull/102.diff">https://git.openjdk.org/jmh/pull/102.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jmh/pull/102#issuecomment-1530990423)

### CODE REVIEW

File: jmh-core/src/main/java/org/openjdk/jmh/runner/WorkerThreadFactories.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2005, 2013, Oracle and/or its affiliates. All rights reserved.
```
Comment (shipilev) [inline jmh-core/src/main/java/org/openjdk/jmh/runner/WorkerThreadFactories.java:2]: Copyright years need to be updated.

File: jmh-core/src/main/java/org/openjdk/jmh/runner/WorkerThreadFactories.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        try {
+            Method m = Class.forName("java.lang.Thread").getMethod("ofVirtual");
+            Object threadBuilder = m.invoke(null);
+            Class<?> threadBuilderClazz = Class.forName("java.lang.Thread$Builder");
+            m = threadBuilderClazz.getMethod("name", String.class, long.class);
+            m.invoke(threadBuilder, prefix + "-jmh-worker-", 1L);
+            m = threadBuilderClazz.getMethod("factory");
+            return (ThreadFactory) m.invoke(threadBuilder);
+        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException |
+                 ClassNotFoundException  | NullPointerException e) {
```
Comment (shipilev) [inline jmh-core/src/main/java/org/openjdk/jmh/runner/WorkerThreadFactories.java:72]:
```suggestion
                 ClassNotFoundException | NullPointerException e) {
```

File: jmh-core/src/main/java/org/openjdk/jmh/runner/WorkerThreadFactories.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            return thread;
+        }
+    }
+
+    static ThreadFactory getPlatformWorkerFactory(String prefix) {
+        return new WorkerThreadFactory(prefix);
+    }
+
+    static ThreadFactory getVirtualWorkerFactory(String prefix) {
+        // requires reflection to be compilable in pre JDK 21
```
Comment (shipilev) [inline jmh-core/src/main/java/org/openjdk/jmh/runner/WorkerThreadFactories.java:62]:
```suggestion
        // This API is only available in JDK 21+. Use reflection to make the code compilable with lower JDKs.
```

File: jmh-core/src/main/java/org/openjdk/jmh/runner/WorkerThreadFactories.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicInteger;
+
+class WorkerThreadFactories {
+
+    static class WorkerThreadFactory implements ThreadFactory {
```
Comment (shipilev) [inline jmh-core/src/main/java/org/openjdk/jmh/runner/WorkerThreadFactories.java:36]: This is `PlatformThreadFactory` now?

File: jmh-core/src/main/java/org/openjdk/jmh/runner/WorkerThreadFactories.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            Method m = Class.forName("java.lang.Thread").getMethod("ofVirtual");
+            Object threadBuilder = m.invoke(null);
+            Class<?> threadBuilderClazz = Class.forName("java.lang.Thread$Builder");
+            m = threadBuilderClazz.getMethod("name", String.class, long.class);
+            m.invoke(threadBuilder, prefix + "-jmh-worker-", 1L);
+            m = threadBuilderClazz.getMethod("factory");
+            return (ThreadFactory) m.invoke(threadBuilder);
+        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException |
+                 ClassNotFoundException  | NullPointerException e) {
+            throw new RuntimeException("Can't instantiate VirtualThreadFactory", e);
```
Comment (shipilev) [inline jmh-core/src/main/java/org/openjdk/jmh/runner/WorkerThreadFactories.java:73]:
```suggestion
            throw new RuntimeException("Cannot instantiate VirtualThreadFactory", e);
```

File: jmh-core/src/main/java/org/openjdk/jmh/runner/WorkerThreadFactories.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        @Override
+        public Thread newThread(Runnable r) {
+            Thread thread = factory.newThread(r);
+            thread.setName(prefix + "-jmh-worker-" + counter.incrementAndGet());
+            thread.setDaemon(true);
+            return thread;
+        }
+    }
+
+    static ThreadFactory getPlatformWorkerFactory(String prefix) {
```
Comment (shipilev) [inline jmh-core/src/main/java/org/openjdk/jmh/runner/WorkerThreadFactories.java:57]: `Workers.platformWorkerFactory` and `Workers.virtualWorkerFactory` might be better namess.

**Review by shipilev — COMMENTED** [review]
Looks okay, some comments:

**Review by shipilev — APPROVED** [review]
Looks fine, provided GHA is green.

At some point in the future, we would need to add a regression test for all these executor types, but today it is not completely possible, because there is no 21-ea Temurin build for GHA yet.

**Comment by kuksenko** [comment]
> Looks fine, provided GHA is green.
> 
> At some point in the future, we would need to add a regression test for all these executor types, but today it is not completely possible, because there is no 21-ea Temurin build for GHA yet.

Will do.

**Comment by shipilev** [comment]
> > At some point in the future, we would need to add a regression test for all these executor types, but today it is not completely possible, because there is no 21-ea Temurin build for GHA yet.
> 
> Will do.

Thanks. Track when #101 is done.

**Comment by shipilev** [comment]
> > > At some point in the future, we would need to add a regression test for all these executor types, but today it is not completely possible, because there is no 21-ea Temurin build for GHA yet.
> > 
> > Will do.
> 
> Thanks. Track when #101 is done.

Feel free to add the executor tests now.

**Pull request closed without merging**